larq_zoo/literature/real_to_bin_nets.py

```diff
@@ -0,0 +1,491 @@
+            use_bias=False,
+        )(x)
+        x = tf.keras.layers.BatchNormalization(
+            momentum=self.momentum, name=f"{name}_batch_norm"
+        )(x)
+
+        # StrongBaselineNet uses PReLU; ResNets and Real-to-Bin nets use ReLU.
+        if use_prelu:
+            x = tf.keras.layers.PReLU(shared_axes=[1, 2], name=f"{name}_prelu")(x)
+        else:
+            x = tf.keras.layers.ReLU(name=f"{name}_relu")(x)
+
+        return tf.keras.layers.MaxPool2D(
+            3, strides=2, padding="same", name=f"{name}_pool"
+        )(x)
+
+    def last_block(self, x, name=""):
+        """Last block, shared across ResNet, StrongBaselineNet and Real-to-Bin nets."""
+
+        x = tf.keras.layers.GlobalAvgPool2D(name=f"{name}_global_pool")(x)
+        x = tf.keras.layers.Dense(
+            self.num_classes, activation=None, name=f"{name}_logits",
+        )(x)
+        return tf.keras.layers.Softmax(name=f"{name}_probs", dtype=tf.float32)(x)
+
+    def block(self, x, downsample=False, name=""):
+        """Main network block
+
+        This block differs between ResNet and StrongBaseline / Real-to-Bin Nets.
+        It is implemented by the ResNet18 and StrongBaselineNet subclasses.
+        """
+
+        raise NotImplementedError()
+
+    def shortcut_connection(
+        self, x: tf.Tensor, name: str, in_channels: int, out_channels: int
+    ):
+        if in_channels == out_channels:
+            return x
+        x = tf.keras.layers.AvgPool2D(
+            2, strides=2, padding="same", name=f"{name}_shortcut_pool"
+        )(x)
+        x = tf.keras.layers.Conv2D(
+            out_channels,
+            kernel_size=1,
+            kernel_initializer=self.kernel_initializer,
+            kernel_regularizer=self.kernel_regularizer,
+            use_bias=False,
+            name=f"{name}_shortcut_conv2d",
+        )(x)
+        return tf.keras.layers.BatchNormalization(
+            momentum=self.momentum, name=f"{name}_shortcut_batch_norm"
+        )(x)
+
+    def build(self) -> tf.keras.models.Model:
+        """Build the model."""
+
+        x = self.first_block(
+            self.image_input,
+            name=f"{self.model_name}_block_1",
+            use_prelu=isinstance(self, StrongBaselineNetFactory),
+        )
+        for block in range(2, 10):
+            x = self.block(
+                x,
+                name=f"{self.model_name}_block_{block}",
+                downsample=block % 2 == 0 and block > 3,
+            )
+
+        if self.include_top:
+            x = self.last_block(x, name=f"{self.model_name}_block_10")
+
+        model = tf.keras.Model(
+            inputs=self.image_input, outputs=x, name=self.model_name,
+        )
+
+        # Load weights.
+        if self.weights == "imagenet":
+            model.load_weights(self._get_weights_path())
+        elif self.weights is not None:
+            model.load_weights(self.weights)
+        return model
+
+    def _get_weights_path(self):
```

Comment:
```suggestion
    @abc.abstractmethod
    def _get_weights_path(self) -> str:
```

Reply:
Also should we maybe make it clear that these are imagenet weights and call it `_get_imagenet_weights_path()`?

Reply:
We actually only provide weights for the BNN version of the R2B net. So this method (which says there are no weights) is the default. It is only overwritten for the R2B subclass and then only provides Imagenet weights.

Reply:
Ah I see, makes sense (for not being an ABC). How do you feel about renaming the function?

Reply:
And perhaps this should be a `ValueError`, because although it's true that the method is not implemented, a user would only see this if they put in the wrong value for the `weights` argument.

Reply:
It is now a `ValueEror` and the name is changed (I forgot that there was an explicit check that the name is `imagenet` before calling this function).